pkg/lib/authn/authenticator/password/generator.go

```diff
@@ -11,13 +90,135 @@ type Generator struct {
 	PwDigitRequired     bool
 	PwSymbolRequired    bool
 	PwMinGuessableLevel int
+	PwExcludedKeywords  []string
 }
 
-type RandSource interface {
+func (g *Generator) Generate() (string, error) {
+	for i := 0; i < MaxTrials; i++ {
+		password, err := g.generate()
+		if err != nil {
+			return "", err
+		}
+
+		if err := g.checkPassword(password); err == nil {
+			return password, nil
+		}
+	}
+
+	return "", ErrPasswordGenerateFailed
 }
 
-type CryptoRandSource struct{}
+func (g *Generator) generate() (string, error) {
+	var charList = g.prepareCharList()
 
-func (g *Generator) Generate() (string, error) {
-	return "Generated", nil
+	// Ensure the password is at least the minimum length.
+	var minLength = g.PwMinLength
+	if minLength < DefaultMinLength {
+		minLength = DefaultMinLength
+	}
+
+	// Override min length if min guessable level is enabled.
+	if g.PwMinGuessableLevel > 0 {
+		minLength = GuessableEnabledMinLength
+	}
+
+	var password strings.Builder
+
+	if g.PwUppercaseRequired {
+		c, err := g.RandSource.Pick(CharListUppercase)
+		if err != nil {
+			return "", err
+		}
+		password.WriteByte(CharListUppercase[c])
+	}
+	if g.PwLowercaseRequired {
+		c, err := g.RandSource.Pick(CharListLowercase)
+		if err != nil {
+			return "", err
+		}
+		password.WriteByte(CharListLowercase[c])
+	}
+	if g.PwAlphabetRequired && !g.PwUppercaseRequired && !g.PwLowercaseRequired {
+		c, err := g.RandSource.Pick(CharListAlphabet)
+		if err != nil {
+			return "", err
+		}
+		password.WriteByte(CharListAlphabet[c])
+	}
+	if g.PwDigitRequired {
+		c, err := g.RandSource.Pick(CharListDigit)
+		if err != nil {
+			return "", err
+		}
+		password.WriteByte(CharListDigit[c])
+	}
+	if g.PwSymbolRequired {
+		c, err := g.RandSource.Pick(CharListSymbol)
+		if err != nil {
+			return "", err
+		}
+		password.WriteByte(CharListSymbol[c])
+	}
+
+	for j := 0; j < minLength; j++ {
+		c, err := g.RandSource.Pick(charList)
+		if err != nil {
+			return "", err
+		}
+		password.WriteByte(charList[c])
+	}
+
+	shuffled, err := g.RandSource.Shuffle(password.String())
+	if err != nil {
+		return "", err
+	}
+
+	return shuffled, nil
+}
+
+func createSetFromList(list string) string {
+	var set strings.Builder
+	setMap := make(map[byte]bool)
+
+	for i := 0; i < len(list); i++ {
+		if _, ok := setMap[list[i]]; !ok {
+			set.WriteByte(list[i])
+			setMap[list[i]] = true
+		}
+	}
+
+	return set.String()
+}
+
+func (g *Generator) prepareCharList() string {
+	charList := DefaultCharList
+
+	if g.PwLowercaseRequired {
+		charList += CharListLowercase
+	}
+	if g.PwUppercaseRequired {
+		charList += CharListUppercase
+	}
+	if g.PwAlphabetRequired {
+		charList += CharListAlphabet
+	}
+	if g.PwDigitRequired {
+		charList += CharListDigit
+	}
+	if g.PwSymbolRequired {
+		charList += CharListSymbol
+	}
+
+	for _, keyword := range g.PwExcludedKeywords {
+		charList = strings.ReplaceAll(charList, keyword, "")
+	}
+
+	// Deduplication
+	charList = createSetFromList(charList)
```

Comment:
Remove createSetFromList, you should be able to create a character set without duplication.

```
type characterSet int

const (
  characterSetLowercase = iota
  ...
)

var allCharacterSets = []characterSet{
  characterSetLowercase,
  // ...
}

func (s characterSet) Append(w io.Writer) {
  switch s {
    case characterSetLowercase:
      w.Write([]byte("abc...."))
  }
}

func (g *Generator) prepareCharList() string {
  set := map[characterSet]struct{}
  if g.PwLowercaseRequired {
    set[characterSetLowercase] = struct{}{}
  }
  // ...
  // Add the default set.
  if _, ok := set[characterSetLowercase]; !ok {
    set[characterSetLowercase] = struct{}{}
  }
  var buf strings.Builder
  for cs := range allCharacterSets {
    if _, ok := set[cs]; ok {
      cs.Apppend(buf)
    }
  }

  return buf.String()
}
```